core/src/main/java/hudson/model/Label.java

```diff
@@ -611,9 +620,34 @@ public static Label get(String l) {
      *
      * TODO: replace this with a real parser later
      */
-    public static Label parseExpression(@NonNull String labelExpression) throws ANTLRException {
-        LabelExpressionLexer lexer = new LabelExpressionLexer(new StringReader(labelExpression));
-        return new LabelExpressionParser(lexer).expr();
+    public static Label parseExpression(@NonNull String labelExpression) throws ParseCancellationException {
+        LabelExpressionLexer lexer = new LabelExpressionLexer(CharStreams.fromString(labelExpression));
+        lexer.removeErrorListeners();
+        lexer.addErrorListener(new ANTLRErrorListener() {
+            @Override
+            public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int line, int charPositionInLine, String msg, RecognitionException e) {
+                throw new ParseCancellationException(msg);
```

Comment:
Here and elsewhere, we need to throw our new subtype of `ParseCancellationException` (`ANTLRException`) in order to provide actual compatibility (not just linker pacification) for callers like https://github.com/jenkinsci/periodicbackup-plugin/blob/92b9ebaa0758333256e3478ea8cff5bc07b28085/src/main/java/org/jenkinsci/plugins/periodicbackup/PeriodicBackup.java#L76-L78 that are expecting an `ANTLRException` to be thrown in order to catch it and do something interesting. Although we expect those callers to eventually migrate to catching `ParseCancellationException` instead (due to the `@deprecation` Javadoc I added in commit c63d928c9efddd45ec411322927ddbc4749568e6), existing callers that were compiled against older cores still need to get the same behavior they are already expecting. It is only after we have made a good faith effort to migrate callers from `ANTLRException` to `ParseCancellationException` (and released those PRs!) that we can consider changing core's behavior to throw something that is not assignable to `ANTLRException`.